### PR TITLE
Implement encode functions for VectorValue and LazyBatchColumn(Vec)

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR implements `encode()` and `maximum_encoded_size()` for `VectorValue`, `LazyBatchColumn` and `LazyBatchColumnVec`.

Extracted from #3898, based on #4208, #4209, #4211, #4212, #4213 and #4214.

* [x]      Merge #4208.
* [x]      Merge #4209.
* [x]      Merge #4211.
* [x]      Merge #4212.
* [x]      Merge #4213.
* [x]      Merge #4214.

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: src/coprocessor/codec/data_type/vector.rs [hunk 1]
```diff
@@ -443,6 +443,160 @@ impl VectorValue {
...
+    /// Returns maximum encoded size in binary format.
+    pub fn maximum_encoded_size(&self) -> Result<usize> {
+        match self {
+            VectorValue::Int(ref vec) => Ok(vec.len() * 9),
+
+            // Some elements might be NULLs which encoded size is 1 byte. However it's fine because
+            // this function only calculates a maximum encoded size (for constructing buffers), not
+            // actual encoded size.
+            VectorValue::Real(ref vec) => Ok(vec.len() * 9),
+            VectorValue::Decimal(ref vec) => {
```
Comment (AndreMouche) [inline src/coprocessor/codec/data_type/vector.rs]: In chunk, we process `Decimal` as a fixed length struct, maybe we could process it in a similar way? And when we support return data encoded by chunk, would we change this function?

File: src/coprocessor/codec/data_type/vector.rs [hunk 2]
```diff
@@ -443,6 +443,160 @@ impl VectorValue {
...
+        use crate::util::codec::number::NumberEncoder;
+
+        match self {
+            VectorValue::Int(ref vec) => {
+                match vec[row_index] {
+                    None => {
+                        output.push(datum::NIL_FLAG);
+                    }
+                    Some(val) => {
+                        output.push(datum::INT_FLAG);
```
Comment (AndreMouche) [inline src/coprocessor/codec/data_type/vector.rs]: Should we consider to mark `u64` with `UINT_FLAG`?

File: src/coprocessor/codec/data_type/vector.rs [hunk 3]
```diff
@@ -443,6 +443,160 @@ impl VectorValue {
...
+                        output.encode_i64(val)?;
```
Comment (AndreMouche) [inline src/coprocessor/codec/data_type/vector.rs]: It seems now in the `master` we will comparable `i64` into `var_i64`.

File: src/coprocessor/codec/data_type/vector.rs [hunk 4]
```diff
@@ -443,6 +443,160 @@ impl VectorValue {
...
+                match &vec[row_index] {
+                    None => {
+                        output.push(datum::NIL_FLAG);
+                    }
+                    Some(ref val) => {
+                        output.push(datum::JSON_FLAG);
+                        output.encode_json(val)?;
+                    }
+                }
+                Ok(())
```
Comment (AndreMouche) [inline src/coprocessor/codec/data_type/vector.rs]: Should we consider `MAX` and `MIN`?

File: src/coprocessor/codec/data_type/vector.rs [hunk 5]
```diff
@@ -443,6 +443,160 @@ impl VectorValue {
 
         Ok(())
     }
+
+    /// Returns maximum encoded size in binary format.
+    pub fn maximum_encoded_size(&self) -> Result<usize> {
+        match self {
+            VectorValue::Int(ref vec) => Ok(vec.len() * 9),
+
+            // Some elements might be NULLs which encoded size is 1 byte. However it's fine because
```
Comment (rleungx) [inline src/coprocessor/codec/data_type/vector.rs]: Will it be a problem if there are almost NULLs?

Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: No. It will just allocate more memory than really used, but will not cause any problems.

File: src/coprocessor/codec/data_type/vector.rs (hunk 1 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: > In chunk, we process `Decimal` as a fixed length struct, maybe we could process it in a similar way?

It depends on the format itself. However we can assign a maximum size (instead of an approximate max size) for it so that the size can be just `vec.len() * N` which is very fast to calculate.

> And when we support return data encoded by chunk, would we change this function?

Yes. We need to change both `maximum_encoded_size()` and `encode()`. However I prefer to directly reuse encoding functions in chunk when we can return data in chunk.

File: src/coprocessor/codec/data_type/vector.rs [hunk 6]
```diff
@@ -443,6 +443,160 @@ impl VectorValue {
...
+            }
+        }
+    }
+
+    /// Encodes a single element into binary format.
+    pub fn encode(&self, row_index: usize, output: &mut Vec<u8>) -> Result<()> {
+        use crate::coprocessor::codec::mysql::DecimalEncoder;
+        use crate::coprocessor::codec::mysql::JsonEncoder;
+        use crate::util::codec::bytes::BytesEncoder;
+        use crate::util::codec::number::NumberEncoder;
```
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: They are only used in this function, so that I prefer not to do it?

File: src/coprocessor/codec/data_type/vector.rs (hunk 2 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: I don't know if it is necessary. Depends on whether or not TiDB supports it. According to integration test it supports. @xuhuaiyu Please confirm: For field type = u64, will TiDB correctly parse non-chunk format (datum format) with INT flag?

File: src/coprocessor/codec/data_type/vector.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: Yes. It utilizes the same feature as above that (looks like) TiDB can correctly parse datum based on the flag in the datum instead of the given field type. (@XuHuaiyu Please help confirm)

The reason that use `i64` instead of `var_i64` is to be efficient.

File: src/coprocessor/codec/data_type/vector.rs (hunk 4 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: There is no MAX and MIN in the VectorValue for now, so for this piece of code we don't need it.

By the way when will we meet a MAX datum or a MIN datum? @Xuhuaiyu PTAL

Comment (XuHuaiyu) [reply src/coprocessor/codec/data_type/vector.rs]: FYI, MAX and MIN datum will be used when building range.

File: src/coprocessor/codec/data_type/vector.rs (hunk 2 above)
Comment (XuHuaiyu) [reply src/coprocessor/codec/data_type/vector.rs]: Maybe this is what you want.
``` go
// AppendDatum appends a datum into the chunk.
func (c *Chunk) AppendDatum(colIdx int, d *types.Datum) {
	switch d.Kind() {
	case types.KindNull:
		c.AppendNull(colIdx)
	case types.KindInt64:
		c.AppendInt64(colIdx, d.GetInt64())
	case types.KindUint64:
		c.AppendUint64(colIdx, d.GetUint64())
	case types.KindFloat32:
		c.AppendFloat32(colIdx, d.GetFloat32())
	case types.KindFloat64:
		c.AppendFloat64(colIdx, d.GetFloat64())
	case types.KindString, types.KindBytes, types.KindBinaryLiteral, types.KindRaw, types.KindMysqlBit:
		c.AppendBytes(colIdx, d.GetBytes())
	case types.KindMysqlDecimal:
		c.AppendMyDecimal(colIdx, d.GetMysqlDecimal())
	case types.KindMysqlDuration:
		c.AppendDuration(colIdx, d.GetMysqlDuration())
	case types.KindMysqlEnum:
		c.AppendEnum(colIdx, d.GetMysqlEnum())
	case types.KindMysqlSet:
		c.AppendSet(colIdx, d.GetMysqlSet())
	case types.KindMysqlTime:
		c.AppendTime(colIdx, d.GetMysqlTime())
	case types.KindMysqlJSON:
		c.AppendJSON(colIdx, d.GetMysqlJSON())
	}
}
```

File: src/coprocessor/codec/data_type/vector.rs (hunk 3 above)
Comment (XuHuaiyu) [reply src/coprocessor/codec/data_type/vector.rs]: > The reason that uses `i64` instead of `var_i64` is to be efficient.

What's the result of the comparison? This may reduce the transmission efficiency in most cases.

File: src/coprocessor/codec/data_type/vector.rs (hunk 4 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: @XuHuaiyu Will MAX and MIN being used during calculation (i.e. the pushed down value from TiDB) or stored in row?

File: src/coprocessor/codec/data_type/vector.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: See https://github.com/tikv/tikv/pull/3629

File: src/coprocessor/codec/data_type/vector.rs (hunk 2 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: But we are not using Chunk?

File: src/coprocessor/codec/data_type/vector.rs (hunk 4 above)
Comment (XuHuaiyu) [reply src/coprocessor/codec/data_type/vector.rs]: no

File: src/coprocessor/codec/data_type/vector.rs (hunk 2 above)
Comment (XuHuaiyu) [reply src/coprocessor/codec/data_type/vector.rs]: I do not really understand what do you mean.
This is a member function of Chunk.

File: src/coprocessor/codec/data_type/vector.rs (hunk 3 above)
Comment (AndreMouche) [reply src/coprocessor/codec/data_type/vector.rs]: How about the benchmark for YCSB? It seems there is a story why we use `varint ` to short the value in network transmission @coocood

File: src/coprocessor/codec/data_type/vector.rs (hunk 4 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: Thanks!

File: src/coprocessor/codec/data_type/vector.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: @AndreMouche As long as you switch to Arrow format later, it will be the same. In Arrow we cost even more on network. I think the main concern here should be whether or not TiDB can correctly accept this behaviour.

File: src/coprocessor/codec/data_type/vector.rs (hunk 2 above)
Comment (breeswish) [reply src/coprocessor/codec/data_type/vector.rs]: @AndreMouche I have changed to encode uint separately. PTAL

File: src/coprocessor/codec/batch/lazy_column.rs [hunk 7]
```diff
@@ -553,7 +598,7 @@ mod benches {
         }
     }
 
-    /// Bench performance of push 10 bytes to a vector based LazyBatchColumn.
+    /// Bench performance of pushing 10 bytes to a vector based LazyBatchColumn.
```
Comment (AndreMouche) [inline src/coprocessor/codec/batch/lazy_column.rs]: Could we move the bench test into the `benches` directory in another PR?

File: src/storage/mod.rs [hunk 8]
```diff
@@ -687,49 +687,45 @@ impl<E: Engine> Storage<E> {
...
-                let ctxd = ctxd.clone();
-                let mut thread_ctx = ctxd.current_thread_context_mut();
-                thread_ctx.start_command_duration_timer(CMD, priority)
-            };
-
-            Self::async_snapshot(engine, &ctx)
-                .and_then(move |snapshot: E::Snap| {
-                    let mut thread_ctx = ctxd.current_thread_context_mut();
-                    let _t_process = thread_ctx.start_processing_read_duration_timer(CMD);
+        let timer = SCHED_HISTOGRAM_VEC
```
Comment (AndreMouche) [inline src/storage/mod.rs]: I think we should do it in another PR.

Comment (breeswish) [reply src/storage/mod.rs]: Oh this is a merge issue.

File: src/coprocessor/codec/batch/lazy_column.rs (hunk 7 above)
Comment (breeswish) [reply src/coprocessor/codec/batch/lazy_column.rs]: But we are not introducing benchmarks in this PR?
